Planning process documentation (#441)
kokkos/kokkos-core-wiki: docs/source/testing-and-issue-tracking/Kokkos-Project-Planning.md

# Kokkos Project Planning

## Requirements Gathering

> There are three requirement categories for the Kokkos Core project:

Reviewer: this says three but below there are 4 bullets

Reviewer 2: ```suggestion
There are four requirement categories for the Kokkos Core project:
```

thus share the maintenance burden with the entire C++ implementer community in the long run.

To enable the on-ramp, Kokkos will provide backports of ISO C++ features to prior C++ standards, where appropriate and desired.
Kokkos will also provide extensions of ISO C++ features which work on GPUs, something which is not available by default.

Kokkos features which have proven themselves, and are of interest to a wide audience are evaluated for possible inclusion in the ISO C++ standard.
The Kokkos team will write proposals for the ISO C++ committee when appropriate.

If a feature is included in the ISO C++ standard, the Kokkos team will make the API variants provided in the future C++ standard
available on currently Kokkos supported software stacks to the greatest extent possible.

#### Activities to support this requirement:

- participate in ISO C++ committee meetings
- monitor requests for ISO C++ features to be provided by Kokkos
- write proposals for ISO C++ for mature Kokkos features with wide applicability
- backport relevant future ISO C++ features to standards supported by Kokkos

## Release Planning

> Kokkos releases are based on the "catch the train" model - i.e. the primary goal is to have regular releases,
> not a specific feature list for each release.
> 
> Major releases happen every three years, minor releases are aimed at every 3-4 months, with additional patch releases as necessary.
> 
> The primary difference between a major and a minor release is that deprecated features are only removed at major releases, and
> major releases come with a bump in minimal compiler version requirements, and an updated minimum ISO C++ standard version.
> Other than that, there is no difference in planning and execution of major and minor releases.
> 
> The difference of patch releases to minor and major releases is that patch releases  will generally only contain bugfixes, and not new capabilities.
> 
> At the beginning of a release cycle the Kokkos Core leadership will determine high priority thrusts for the release cycle.
> Furthermore, each team member will make a list of their personal priorities for the release cycle.
> The priorities are discussed and refined at the Kokkos developer meeting and collected in internal documents.
> 
> Issues for each item are assigned to the [Kokkos project plan](https://github.com/orgs/kokkos/projects/1) including team member assignments.
> 
> The [Kokkos project plan](https://github.com/orgs/kokkos/projects/1) assigns issues to one of 7 categories:
> 
> - *Unassigned:* issues that aren't assigned yet to team members.
> - *Unassigned - Priority:* issues that aren't assigned yet to team members, but are high priority. These should be assigned at the next weekly developer meeting.
> - *To Do:* Issue was assigned to a team member, but is not yet actively worked on.
> - *To Do - Priority:* Issue was assigned to a team member, but is not yet actively worked on. It is expected to be the next item in the queue of the assigned developer. If this item does not transition to *In Progress* by the next developer meeting, reassignment is considered.
> - *In Progress:* Issue is getting worked on.
> - *In Progress - Priority:* Issue is getting worked on. Code reviews for this issue are considered priority, in order to get this merged as soon as possible.
> - *Done:* Issue is addressed via merged pull request, or was closed because of new information which made it obsolete. For merged pull requests it is ensured that a Changelog entry was generated if appropriate, before removing the item from the project plan.

Reviewer 3: Note that a lot of (most of?) the items in the planning board are pull requests, not issues. This is particularly confusing for new pull requests that are "Unassigned" by default but probably should rather be "In Progress".

```suggestion
Kokkos releases are based on the "catch the train" model - i.e. the primary goal is to have regular releases,
not a specific feature list for each release.

Major releases happen every three years, minor releases are aimed at every 3-4 months, with additional patch releases as necessary.

The primary difference between a major and a minor release is that deprecated features are only removed at major releases, and
major releases come with a bump in minimal compiler version requirements and an updated minimum ISO C++ standard version.
Other than that, there is no difference in the planning and execution of major and minor releases.

In contrast to major and minor releases, patch releases generally only contain bug fixes and no new capabilities.

At the beginning of a release cycle, the Kokkos Core leadership will determine high-priority thrusts for the release cycle.
Furthermore, each team member will make a list of their personal priorities for the release cycle.
The priorities are discussed and refined at the Kokkos developer meeting and collected in internal documents.

Issues for each item are assigned to the [Kokkos project plan](https://github.com/orgs/kokkos/projects/1) including team member assignments.

The [Kokkos project plan](https://github.com/orgs/kokkos/projects/1) assigns issues to one of 7 categories:

- *Unassigned:* issues that aren't assigned yet to team members.
- *Unassigned - Priority:* issues that aren't assigned yet to team members, but are high priority. These should be assigned at the next weekly developer meeting.
- *To Do:* Issue was assigned to a team member but is not yet actively worked on.
- *To Do - Priority:* Issue was assigned to a team member, but is not yet actively worked on. It is expected to be the next item in the queue of the assigned developer. If this item does not transition to *In Progress* by the next developer meeting, reassignment is considered.
- *In Progress:* Issue is getting worked on.
- *In Progress - Priority:* Issue is getting worked on. Code reviews for this issue are considered a priority, in order to get this resolved as soon as possible.
- *Done:* Issue is addressed via merged pull request, or was closed because of new information which made it obsolete. For merged pull requests it is ensured that a changelog entry was generated, if appropriate, before removing the item from the project plan.
```
Issues don't get "merged".

They can ask questions about parts of the code they are less familiar with or
ask for feedback on any ongoing issue.
Conversations on Slack are to be considered as ephemeral.  Messages older than 90 days are deleted (unpaid plan).
If something needs to be referenceable longer term, then it needs to be discussed on GitHub wherever appropriate.
Private information may be hosted on the [internal repository](https://github.com/kokkos/internal-documents) but do not post NDA data on there.

Kokkos developer meeting held once a week on Wednesdays 2pm ET / 12 pm MT / 18:00 UTC on Zoom.
The agenda is posted on the internal repository ahead of time (it can be found under the [`meeting-notes/`](https://github.com/kokkos/internal-documents/tree/master/meeting-notes/2023) directory).
Developers are allowed to edit the agenda and add topics or issues that they would like to be discussed at the meeting.

## Release Process

The release process has six steps:

- create release candidate branch
- perform integration tests with release candidate
- resolve issues and cherry-pick fixes to release candidate
- check Changelog
- tag a release

> - conduct release briefing for user community

Reviewer 3: https://github.com/kokkos/internal-documents/blob/master/testing/Kokkos_Release_Process.md has 7 steps should these match?

> When nearing a desired release date, the release candidate branch will created from the Kokkos develop branch.
> Before creating the release candidate, possible delay reasons will be discussed at the developer meeting.
> This could include important bug fixes, or an important feature being in the last phase of code review,
> but is generally done under exceptional circumstances.
> Furthermore, major new features may be delayed for merging into the development branch, until after the creation
> of the release candidate.
> This ensures that major new features have a period of testing in the develop branch, before they are shipped.

Reviewer 3: ```suggestion
When nearing a desired release date, the release candidate branch will be created from the Kokkos develop branch.
Before creating the release candidate, possible delay reasons will be discussed at the developer meeting.
This could include important bug fixes, or an important feature being in the last phase of code review,
but is generally done under exceptional circumstances.
Furthermore, merging major new features into the development branch may be delayed until after the creation
of the release candidate.
This ensures that major new features have a period of testing in the develop branch before they are shipped.
```

> After creating the release candidate branch integration testing is started.
> This includes internal testing by the Kokkos team with selected customer codes, as well as partnering
> with some primary customers which will try the release candidate in their testing processes.

Reviewer 3: ```suggestion
After creating the release candidate branch integration testing is started.
This includes internal testing by the Kokkos team with selected customer codes, as well as partnering
with some primary customers who will try the release candidate in their testing processes.
```
